doc/developer/design/20231216_pg_source_definite_error_status.md

```diff
@@ -0,0 +1,79 @@
+
+PG sources never receive definite errors from the perspective of source
+statuses, e.g. if a publication gets dropped. This makes monitoring source
+health very odd––all of the subsources report the definite error in their
+statuses, but the primary source does not.
+
+This is a problem as of
+[#23746](https://github.com/MaterializeInc/materialize/pull/23746), which
+previously papered over this issue by duplicating definite errors in subsources
+to primary sources––however, this document tackles the underlying issue.
+
+This issue will also be more apparent if we ever decide to implement a version
+of [sql: change SUBSOURCES to be a flavor of tables
+#20208](https://github.com/MaterializeInc/materialize/issues/20208) because the
+definite errors could then totally disappear from the source status table.
+
+## Success Criteria
+
+Definite errors get propagated to the source status table for the primary PG
+source.
+
+## Out of Scope
+
+This issue is narrowly scoped enough that there is nothing out of scope.
+
+## Solution Proposal
+
+The fundamental issue causing this behavior is a confluence of:
+
+-   We never want to produce data for the PG source's primary collection (as we do
+    for subsources), as it is "useless" and we eventually want to remove it
+    entirely.
+-   We currently and reasonably rely on the dataflows' health output to derive
+    statuses.
+-   `build_fallible`, which runs our fallible async operators, only produces
+    health messages when the dataflow errors––however, when the dataflow errors,
+    we place it in a "suspended" state, awaiting to be restarted. This suspended
+    state is very careful in its handling of the dataflow's output frontiers,
+    ensuring they do no emit empty frontiers when dropped.
+
+    This means that we do not have a "native" mechanism to derive statuses when
+    dataflows exit with an `Ok` result--i.e. we want to exit the dataflow and
+    close all of the input capabilities.
+
+To resolve this, I propose we allow `build_fallible`'s "health" stream to return
+`Result<O, E>` instead of only `E`.
+
+We can then plumb through the definite errors of PG sources in the `O` side of
+the result, e.g. `Result<Result<(), DefiniteError>, TransientError>`. This
+mirrors the API of a number of internal functions in PG sources.
+
+Additionally, this design also creates a clear path to emitting periodic status
+messages in the operator's healthy state––e.g. emitting snapshot progress
+messages via the status table.
+
+## Minimal Viable Prototype
+
+See provided commits.
+
+## Alternatives
+
+It is possible to "handroll" this same behavior by:
+
+-   Continuing to only produce health messages on `build_fallible`'s "error
+    channel."
+-   Explicitly dropping all of the inputs whenever returning a definite error.
+
+However, this has the downside of requiring all code paths returning definite
+errors to remember to advance all frontiers. This seems easy to overlook or
```

Comment:
This happens already automatically when you return from a fallible operator, right? What is the additional thing that we'd have to remember?